storage/index.go

```diff
@@ -265,16 +264,53 @@ func (i *nsIndex) Query(
 		return index.QueryResults{}, err
 	}
 
-	idxIter := index.NewIterator(i.nsID, iter, i.opts)
+	var (
+		results    = i.opts.ResultsPool().Get()
+		size       = results.Size()
+		brokeEarly = false
+		success    = false
+	)
+	results.Reset(i.nsID)
+	defer func() {
+		if !success {
+			results.Finalize()
+			exec.Close()
+			iter.Close()
+		}
+	}()
+
+	for iter.Next() {
+		// FOLLOWUP(prateek): as part of the runtime options wiring, also set a server side max limit which
+		// super-seeds the user requested limit. Paranoia can be a good thing.
+		if opts.Limit > 0 && size >= opts.Limit {
+			brokeEarly = true
+			break
+		}
+		d := iter.Current()
+		_, size, err = results.Add(d)
+		if err != nil {
+			return index.QueryResults{}, err
+		}
+	}
 
-	ctx.RegisterFinalizer(resource.FinalizerFn(func() {
-		idxIter.Finalize()
-		exec.Close()
-	}))
+	if err := iter.Err(); err != nil {
+		return index.QueryResults{}, err
+	}
+
+	if err := iter.Close(); err != nil {
+		return index.QueryResults{}, err
+	}
+
+	if err := exec.Close(); err != nil {
```

Comment:
Hm is it safe to double close `exec` and `iter` this (with the defer at the top?)

Maybe better to:

Reply:
It's safe right now but yeah that's cleaner. will change

Reply:
talked offline, ended up with `safeCloser` struct idiom

Reply:
done